config_file/parameters.md
# Parameters
All parameters match the mode of **argparse**, and commonly used parameters are **bold**.
 
## training and validation

Param                       | Function                          | option
:-                          | :-                                | :-
 **cover_train_path**       | path of cover train files         | -
 **cover_valid_path**       | path of cover validation files    | -
 **stego_train_path**       | path of stego train files         | -
 **stego_valid_path**       | path of stego validation files    | -
 **models_path**            | path of models including checkpoint, data, index and meta files   | -
 **logs_path**              | path of log files including train and validation                  | -
 gpu_selection              | mode of gpu selection             | auto(default), manual
 **mode**                   | mode of running                   | train(default), test and steganalysis
 **submode**                | submode of steganalysis           | one and batch(default)
 **carrier**                | carrier for steganalysis          | audio(default) and image
 **network**                | name of designed network          | -
 **batch_size_train**       | batch size for training           | 64(default)
 **batch_size_validation**  | batch size for validation         | 16(default)
  **learing_rate**          | initialized learing rate          | 1e-3(default)
 **epoch**                  | epoch of training                 | 500(default)
 seed                       | seed of random generation         | 1(defalut)
 **is_regulation**          | whether add regulation or not     | True(fault), false
 coeff_regulation           | coefficient of regulation         | 1e-3(default)
 loss_method                | method of loss funcion            | sigmoid_cross_entropy, softmax_cross_entropy, sparse_softmax_cross_entropy(default)
 class_num                  | number of classifier              | 2(default)
 **height**                 | height of input data matrix       | -
 **width**                  | width of input data matrix        | -
 **channel**                | channel of input data matrix      | 1(default), 3
 is_abs                     | whether make absolute or not      | True and False(default)
 is_trunc                   | whether make truncation or not    | True and False(default)
 is_diff                    | whether make difference or not    | True and False(default)
 is_diff_abs                | whether make diff and abs or not  | True and False(default)
 is_abs_diff                | whether make abs and diff or not  | True and False(default)
 threshold                  | threshold of truncation           | True and False(default)
 order                      | order of difference               | True and False(default)
 direction                  | direction of diff(0-row, 1-col)   | True and False(default)
 decay_method               | method of learning rate decay     | fixed, step, exponential(defaut), inverse_time, natural_exp, polynomial
 decay_step                 | step for learning rate decay      | 5000(default)
 decay_rate                 | rate of learing rate decay        | 0.9(default)
 staircase                  | whether the decay the learning rate at discrete intervals or not | True and False(default)
 max_to_keep                | number of models to be saved      | 3(defalut)

## test

Param                       | Function                          | option
:-                          | :-                                | :-
 **cover_test_path**        | path of cover train files         | -
 **stego_test_path**        | path of stego test files          | -
 **models_path**            | path of models including checkpoint, data, index and meta files   | -
 gpu_selection              | mode of gpu selection             | auto(default), manual
 **mode**                   | mode of running                   | train(default), test and steganalysis
 **submode**                | submode of steganalysis           | one and batch(default)
 **carrier**                | carrier for steganalysis          | audio(default) and image
 **network**                | name of designed network          | -
 class_num                  | number of classifier              | 2(default)
 **height**                 | height of input data matrix       | -
 **width**                  | width of input data matrix        | -
 **channel**                | channel of input data matrix      | 1(default), 3
 is_abs                     | whether make absolute or not      | True and False(default)
 is_trunc                   | whether make truncation or not    | True and False(default)
 is_diff                    | whether make difference or not    | True and False(default)
 is_diff_abs                | whether make diff and abs or not  | True and False(default)
 is_abs_diff                | whether make abs and diff or not  | True and False(default)
 threshold                  | threshold of truncation           | True and False(default)
 order                      | order of difference               | True and False(default)
 direction                  | direction of diff(0-row, 1-col)   | True and False(default)

## steganalysis

Param                       | Function                          | option
:-                          | :-                                | :-
 **files_path**             | path of files for steganalysis    | -
 **models_path**            | path of models including checkpoint, data, index and meta files   | -
 gpu_selection              | mode of gpu selection             | auto(default), manual
 **mode**                   | mode of running                   | train(default), test and steganalysis
 **submode**                | submode of steganalysis           | one and batch(default)
 **carrier**                | carrier for steganalysis          | audio(default) and image
 **network**                | name of designed network          | -
 class_num                  | number of classifier              | 2(default)
 **height**                 | height of input data matrix       | -
 **width**                  | width of input data matrix        | -
 **channel**                | channel of input data matrix      | 1(default), 3
 is_abs                     | whether make absolute or not      | True and False(default)
 is_trunc                   | whether make truncation or not    | True and False(default)
 is_diff                    | whether make difference or not    | True and False(default)
 is_diff_abs                | whether make diff and abs or not  | True and False(default)
 is_abs_diff                | whether make abs and diff or not  | True and False(default)
 threshold                  | threshold of truncation           | True and False(default)
 order                      | order of difference               | True and False(default)
 direction                  | direction of diff(0-row, 1-col)   | True and False(default)
 

config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Created on 2018.01.05
Finished on 2018.01.05
@author: Wang Yuntao
"""

import os
import argparse
from utils import fullfile, get_files_list


"""
function:
        command_parse()                                     parse command line parameters (命令行解析)
        file_path_setup(args)                               get file path (获取文件路径)
"""

"""
    在tensorflow的log日志等级如下： 
    - 0：显示所有日志（默认等级） 
    - 1：显示info、warning和error日志 
    - 2：显示warning和error信息 
    - 3：显示error日志信息 
"""
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
needed_packages = ["tensorflow-gpu", "numpy", "matplotlib", "pandas", "scikit-image", "scikit-learn"]
system = platform.system()


def command_parse():
    """
    :param: NULL
    :return:
        argument
    """
    parser = argparse.ArgumentParser(description="Audio/Image steganalysis with CNN based on tensorflow.")
    print(parser.description)

    # get the current path
    current_path = os.getcwd()
    print("current path: %s" % current_path)

    # mode
    parser.add_argument("--gpu_selection", type=str, default="auto",
                        help="GPU selection mode, if \"auto\", no serial number is needed, otherwise appoint the serial number "
                             "(default: auto, other choice is manual)")
    parser.add_argument("--gpu", type=str, default="0", help="the index of GPU")
    parser.add_argument("--mode", type=str, default="train", help="run mode -- train | test (default: train)")
    parser.add_argument("--submode", type=str, default="one", help="one | batch (default one)")
    parser.add_argument("--carrier", type=str, default="audio", help="image | audio (default audio)")
    parser.add_argument("--network", type=str, default="network1", help="the index of the network (default: network1), "
                                                                        "the detailed introduction of each network is in readme")

    # data info
    parser.add_argument("--data_dir", type=str, help="data set path")
    parser.add_argument("--bitrate", type=int, default=128, help="the bitrate of MP3 files (default:128)")
    parser.add_argument("--relative_payload", type=str, default="2",
                        help="2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10, if the steganography algorithm is the EECS algorithm; 1 | 3 | 5 | 10, otherwise")
    parser.add_argument("--stego_method", type=str, default="EECS",
                        help="EECS | HCM-Gao | HCM- Yan (default: EECS)")
    parser.add_argument("--height", type=int, default=200, help="the height of the QMDCT matrix (default: 200)")
    parser.add_argument("--width", type=int, default=576, help="the width of the QMDCT matrix (default: 576)")
    parser.add_argument("--channel", type=int, default=1, help="the channel of the QMDCT matrix (default: 1)")
    parser.add_argument("--start_index_train", type=int, default=0, help="the start index of file in train folders (default: 0)")
    parser.add_argument("--end_index_train", type=int, default=16000, help="the end index of file in train folders (default: 16000)")
    parser.add_argument("--start_index_valid", type=int, default=0, help="the start index of file in valid folders (default: 0)")
    parser.add_argument("--end_index_valid", type=int, default=4000, help="the end index of file in valid folders (default: 4000)")
    parser.add_argument("--model_file_name", type=str, default="audio_steganalysis", help="model file name (default: audio_steganalysis)")

    # test
    parser.add_argument("--model_files_path", type=str, default=None, help="the model file path used for test (default is None)")
    parser.add_argument("--model_file_path", type=str, default=None, help="the model file path used for test (default is None)")
    parser.add_argument("--steganalysis_file_path", type=str, help="the file path used for steganalysis")
    parser.add_argument("--label", type=str, default=None, help="the label of tested file (default is None)")
    parser.add_argument("--steganalysis_files_path", type=str, help="the files folder path used for steganalysis")
    parser.add_argument("--label_file_path", type=str, help="the label file path used for test")

    # path
    parser.add_argument("--cover_train_path", type=str,
                        help="the path of directory containing cover files for train")
    parser.add_argument("--cover_valid_path", type=str,
                        help="the path of directory containing cover files for valid")
    parser.add_argument("--cover_test_path", type=str,
                        help="the path of directory containing cover files for test")
    parser.add_argument("--stego_train_path", type=str,
                        help="the path of directory containing stego files for train")
    parser.add_argument("--stego_valid_path", type=str,
                        help="the path of directory containing stego files for valid")
    parser.add_argument("--stego_test_path", type=str,
                        help="the path of directory containing stego files for test")
    parser.add_argument("--models_path", type=str, default=current_path + "/models",
                        help="the path of directory containing models")
    parser.add_argument("--logs_path", type=str, default=current_path + "/logs",
                        help="the path of directory containing logs")

    # hyper parameters
    parser.add_argument("--batch_size", type=int, default=128, help="batch size (default: 128 (64 cover|stego pairs))")
    parser.add_argument("--batch_size_train", type=int, default=64, help="batch size for train (default: 64 (32 cover/stego pairs))")
    parser.add_argument("--batch_size_valid", type=int, default=64, help="batch size for valid (default: 64 (32 cover/stego pairs))")
    parser.add_argument("--batch_size_test", type=int, default=64, help="batch size for test (default: 64 (32 cover/stego pairs))")
    parser.add_argument("--learning_rate", type=float, default=1e-3, help="the value of initialized learning rate (default: 1e-3 (0.001))")
    parser.add_argument("--seed", type=int, default=1, help="random seed (default: 1)")
    parser.add_argument("--is_regulation", type=bool, default=True, help="whether regulation or not (default: True)")
    parser.add_argument("--coeff_regulation", type=float, default=1e-3, help="the gain of regulation (default: 1e-3)")
    parser.add_argument("--loss_method", type=bool, default="sparse_softmax_cross_entropy",
                        help="the method of loss calculation (default: sparse_softmax_cross_entropy)")
    parser.add_argument("--class_num", type=int, default=2, help="the class number (default: 2)")
    parser.add_argument("--epoch", type=int, default=500, help="the number of epochs for network training stop (default: 500)")

    # learning rate parameters
    parser.add_argument("--decay_method", type=str, default="exponential", help="the method for learning rate decay (default: exponential)")
    parser.add_argument("--decay_step", type=int, default=5000, help="the step for learning rate decay (default: 5000)")
    parser.add_argument("--decay_rate", type=float, default=0.9, help="the rate for learning rate decay (default: 0.95)")
    parser.add_argument("--staircase", type=bool, default=False,
                        help="whether the decay the learning rate at discrete intervals or not (default:False)")

    # model
    parser.add_argument("--max_to_keep", type=int, default=3, help="the number needed to be saved (default: 3)")
    parser.add_argument("--keep_checkpoint_every_n_hours", type=float, default=0.5,
                        help="how often to keep checkpoints (default: 0.5)")

    # pre-processing
    parser.add_argument("--is_abs", type=bool, default=False, help="abs or not (default: False)")
    parser.add_argument("--is_trunc", type=bool, default=False, help="truncation or not (default: False)")
    parser.add_argument("--threshold", type=int, default=15, help="threshold (default: 15)")
    parser.add_argument("--is_diff", type=bool, default=False, help="threshold (default: False)")
    parser.add_argument("--order", type=int, default=2, help="the order of the difference (default: 2)")
    parser.add_argument("--direction", type=int, default=0, help="0 - row, 1 - col (default: 0)")
    parser.add_argument("--is_diff_abs", type=bool, default=False, help="abs or not after difference (default: False)")
    parser.add_argument("--downsampling", type=bool, default=False, help="downsampling or not (default: False)")
    parser.add_argument("--block", type=int, default=2, help=" (default: 2)")

    arguments = parser.parse_args()

    return arguments


def file_path_setup(args):
    """
    :param args: command-line arguments
    :return:
        cover_train_files_path: the folder of cover files for train
        cover_valid_files_path: the folder of cover files for valid
        stego_train_files_path: the folder of stego files for train
        stego_valid_files_path: the folder of stego files for valid
        model_file_path       : the folder of model files
    """
    # default
    data_dir = args.data_dir                                            # data dir, e.g. E:\Myself\2.database\4.stego
    bitrate = args.bitrate                                              # bitrate, e.g. 128
    stego_method = args.stego_method                                    # stego_method, e.g. EECS
    relative_payload = args.relative_payload                            # relative payload, e.g. 2
    models_path = args.models_path                                      # models path
    logs_path = args.logs_path                                          # logs path

    if stego_method == "EECS":
        stego_dir_name = str(bitrate) + "_W_" + relative_payload + "_H_7_ER_10"
    else:
        if int(relative_payload) < 10:
            stego_dir_name = str(bitrate) + "_0" + relative_payload
        else:
            stego_dir_name = str(bitrate) + "_" + relative_payload

    if data_dir is not None:
        cover_train_files_path = data_dir + "/" + "cover/" + str(bitrate) + "/" + "train"
        cover_valid_files_path = data_dir + "/" + "cover/" + str(bitrate) + "/" + "valid"
        stego_train_files_path = data_dir + "/stego/" + stego_method + "/" + stego_dir_name + "/" + "train"
        stego_valid_files_path = data_dir + "/stego/" + stego_method + "/" + stego_dir_name + "/" + "valid"
        model_file_path = models_path + "/" + stego_dir_name
        log_file_path = logs_path + "/" + stego_dir_name
    else:
        cover_train_files_path = args.cover_train_path
        cover_valid_files_path = args.cover_valid_path
        stego_train_files_path = args.stego_train_path
        stego_valid_files_path = args.stego_valid_path
        model_file_path = args.models_path
        log_file_path = args.logs_path

    # make dir
    if os.path.exists(model_file_path) is False:
        os.mkdir(model_file_path)
    if os.path.exists(log_file_path) is False:
        os.mkdir(log_file_path)

    return cover_train_files_path, cover_valid_files_path, stego_train_files_path, stego_valid_files_path, model_file_path, log_file_path
